### ISSUE TITLE
Apply SPIB in larger system

### ISSUE THREAD

**Issue by YangChemE** [shams-mehdi/SPIB-plumed-masterclass-2022#1]
Hi Shams, 

Not a real issue but a general question. 
I am interested in applying SPIB in larger system like a full kinase domain (~300 - 400 residues), I followed the tutorial and basically implemented the same setting to my protein, then error occurred that system is looking for values outside the grid. I know changing or removing the grid limit may help with this issue, but since the cv used for metad is from an autoencoder (I am not an expert in neural networks), I don't really have a feeling about the latent cv. So I would like to ask if you have any general suggestions for larger system? 

Thanks so much in advance!

Yang

**Comment by shams-mehdi** [comment]
Hello Yang,

Its great to hear that you are checking our SPIB. In the SPIB-plumed masterclass tutorial we provide a heuristic to set up the metadynamics grids. One way to set the grid would be to take the lower dimensional representation from SPIB trained from an input data (let's say the reaction coordinate is 2-d) and calculate minimum, maximum, and standard deviation of this latent data. Then subtract some multiple of the standard deviation from the minimum rc values to get grid min and add some multiple of the standard deviation from the maximum rc values to get grid max.  The python code will look something like the following:

```
rc1_sigma = np.std(traj_rc[:,0])
rc2_sigma = np.std(traj_rc[:,1])
grid_mul = 10

## Set grid_min and grid_max
rc1_grid_min = np.min(traj_rc[:,0]) - grid_mul*rc1_sigma
rc2_grid_min = np.min(traj_rc[:,1]) - grid_mul*rc2_sigma

rc1_grid_max = np.max(traj_rc[:,0]) + grid_mul*rc1_sigma
rc2_grid_max = np.max(traj_rc[:,1]) + grid_mul*rc2_sigma
```

**Comment by YangChemE** [comment]
Hi Shams, 

Thanks for the reply. I did use the code you mentioned the estimate the range of the latent cv, but it still gave me some errors about the grid boundaries. 

I guess I have a more general question (forgive me if it sounds silly), I  guess a reasonable estimation of the grid range would be based on a relatively good trajectory (i.e a trajectory that at least visited some configuration that corresponds to some extreme values of the latent cv, tho convergance is not necessary) that's fed to the neural network for training, am I right? 

Would a initial unbiased simulation with higher temperature (so the trajectory would visit more configurations) improve this situation? 

Thanks, 

Yang

**Comment by shams-mehdi** [comment]
Hi Yang,

It's a very relevant question, thanks for asking!

You are absolutely correct, a good estimate of grid range could be obtained from a relatively good trajectory. As you mentioned you can do the following: (1) train SPIB model from MD trajectory at a particular temperature where the system does not explore much, (2) Perform an unbiased simulation at much higher temperature where the system explores a lot of conformations, (3) Then pass dataset from step (2) through the SPIB model obtained in step (1) from which you can estimate a good grid range.

Although, I would also recommend to simply increase the the grid range by yourself, metadynamics result will not get affected too much by a choice of grid range (e.g, if simulations crashing for grid range between [-2,2], simply make it [-20,20]). You can even continue a crashed simulation from the checkpoint file (.cpt) due to RC going outside of grid range by passing a plumed file with updated/higher grid range.

To summarize, I would not spend too much time to find the "best" grid range. Just choose a reasonable value for which simulation does not crash (if the grid range is unreasonably high then your results will not be as accurate). Finally, since your simulations are crashing because the system is exploring regions in the latent space not observed in the training data, this is actually a good news! Keep at it and please if you have more questions, dont hesitate to ask.

**Comment by YangChemE** [comment]
Hi Shams, Thanks for clarifying!

I got a new question here. I used same way as in the notebook to implement the encoder for the metad (ie, explicitly write down all the calculations step by step in the plumed input file)，however the simulation speed became extremely slow, and the GPU usage is not as much, which makes sense since I'm dealing with a large system where the input OPs is of high dimension (ie, all the backbone dihedrals) . 

I noticed that there are pytorch integrated version of plumed released couple of months ago, which takes the torch.jit saved model for the implementation of the neural networks. I tried that and was expecting the performace to improve as I believe the torch framework should do the nn faster than the words, but it didn't. 

I'm wondering if the torch integrated plumed is indeed not supposed to have much better performance due to maybe the software communication or something, or it is slow because the way I install/use it is not 100% correct.

Such a long question, hope it's not too long, just want to here some expert's opinion. Thanks a lot in advance!

**Comment by shams-mehdi** [comment]
Hi Yang,

Apologies for the delayed response and great questions! Yes, you are correct if the input OPs are really high-dimensional and the encoder is non-linear then even using the torch libraries you will probably see slow simulation speed. We view this as an implementation/PLUMED issue and not directly SPIB issue.

One solution here would be to try linear encoder.  However, if you are already using linear or don't want to use linear encoders, then one way is to use SPIB as a feature selector to reduce number of OPs for metadynamics. To do this, you will have to do two rounds of SPIB. In the first round, you do SPIB with all OPs but use linear encoder (make sure training data has been normalized). Then use the weights of the linear encoder to rank the OPs and identify the most relevant 'x' number of OPs (e.g, x=20). Now train a second round SPIB model using just these 20 features.

Regards,
Shams
